File: system-test/java/src/main/java/com/optivem/shop/dsl/core/usecase/external/clock/usecases/GetTimeVerification.java
package com.optivem.shop.dsl.core.usecase.external.clock.usecases;

import com.optivem.shop.dsl.driver.port.external.clock.dtos.GetTimeResponse;
import com.optivem.shop.dsl.core.shared.ResponseVerification;
import com.optivem.shop.dsl.core.shared.UseCaseContext;
import com.optivem.shop.dsl.common.Converter;

import java.time.Instant;

import static org.assertj.core.api.Assertions.assertThat;

public class GetTimeVerification extends ResponseVerification<GetTimeResponse> {
    public GetTimeVerification(GetTimeResponse response, UseCaseContext context) {
        super(response, context);
    }

    public GetTimeVerification timeIsNotNull() {
        assertThat(getResponse().getTime())
                .withFailMessage("Expected time to be not null")
                .isNotNull();
        return this;
    }

    public GetTimeVerification time(Instant expectedTime) {
        assertThat(getResponse().getTime())
                .withFailMessage("Expected time %s to be equal to %s", getResponse().getTime(), expectedTime)
                .isEqualTo(expectedTime);
        return this;
    }

    public GetTimeVerification time(String expectedTime) {
        return time(Converter.toInstant(expectedTime));
    }

    public GetTimeVerification timeIsAfter(Instant time) {
        assertThat(getResponse().getTime())
                .withFailMessage("Expected time %s to be after %s", getResponse().getTime(), time)
                .isAfter(time);
        return this;
    }

    public GetTimeVerification timeIsBefore(Instant time) {
        assertThat(getResponse().getTime())
                .withFailMessage("Expected time %s to be before %s", getResponse().getTime(), time)
                .isBefore(time);
        return this;
    }

    public GetTimeVerification timeIsBetween(Instant start, Instant end) {
        assertThat(getResponse().getTime())
                .withFailMessage("Expected time %s to be between %s and %s", getResponse().getTime(), start, end)
                .isBetween(start, end);
        return this;
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/clock/usecases/GetTimeVerification.ts
import { expect } from '@playwright/test';
import type { GetTimeResponse } from '../../../../../../driver/port/external/clock/dtos/GetTimeResponse.js';
import { ResponseVerification } from '../../../../shared/response-verification.js';
import type { UseCaseContext } from '../../../../shared/use-case-context.js';

export class GetTimeVerification extends ResponseVerification<GetTimeResponse> {
  constructor(response: GetTimeResponse, context: UseCaseContext) {
    super(response, context);
  }

  timeIsNotNull(): this {
    expect(this.getResponse().time).not.toBeNull();
    expect(this.getResponse().time).not.toBeUndefined();
    return this;
  }

  time(expectedTime: string): this {
    expect(this.getResponse().time).toBe(expectedTime);
    return this;
  }

  timeIsAfter(time: string): this {
    expect(new Date(this.getResponse().time).getTime()).toBeGreaterThan(new Date(time).getTime());
    return this;
  }

  timeIsBefore(time: string): this {
    expect(new Date(this.getResponse().time).getTime()).toBeLessThan(new Date(time).getTime());
    return this;
  }

  timeIsBetween(start: string, end: string): this {
    const actual = new Date(this.getResponse().time).getTime();
    expect(actual).toBeGreaterThanOrEqual(new Date(start).getTime());
    expect(actual).toBeLessThanOrEqual(new Date(end).getTime());
    return this;
  }
}
